src/Serializer/JsonSerializer.php

```diff
@@ -5,48 +5,173 @@
 namespace Patchlevel\EventSourcing\Serializer;
 
 use JsonException;
+use Patchlevel\EventSourcing\Metadata\Event\AttributeEventClassLoader;
+use Patchlevel\EventSourcing\Metadata\Event\AttributeEventMetadataFactory;
+use Patchlevel\EventSourcing\Metadata\Event\EventMetadataFactory;
+use ReflectionClass;
+use TypeError;
 
+use function array_flip;
+use function array_key_exists;
+use function assert;
 use function json_decode;
 use function json_encode;
 
+use const JSON_PRETTY_PRINT;
 use const JSON_THROW_ON_ERROR;
 
 final class JsonSerializer implements Serializer
 {
-    private Hydrator $hydrator;
+    private EventMetadataFactory $metadataFactory;
 
-    public function __construct(?Hydrator $hydrator = null)
+    /** @var array<string, class-string> */
+    private array $eventClassMap;
+
+    /** @var array<class-string, string> */
+    private array $eventClassMapRevert;
+
+    /** @var array<class-string, ReflectionClass> */
+    private array $reflectionClassCache = [];
+
+    /**
+     * @param array<string, class-string> $eventClassMap
+     */
+    public function __construct(EventMetadataFactory $metadataFactory, array $eventClassMap = [])
```

Comment:
I think we should add a EventRegistry class for the events since checking if a class is registered or not should not be the serializers concern. So the code below would look like this:
```php
$eventName = $this->eventRegistry->getByClassName($event::class);
```

instead of 
```php
       if (!array_key_exists($event::class, $this->eventClassMapRevert)) {
            throw new EventClassNotRegistered($event::class);
        }

        $eventName = $this->eventClassMapRevert[$event::class];
```

And also adding `$eventRegistry->getByEventName(): string`

Wdyt?